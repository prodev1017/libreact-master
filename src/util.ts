import * as React from 'react';

export type TComponent<TProps> = React.ComponentClass<TProps> | React.StatelessComponent<TProps>;
export type THoc<P1, P2> = (Comp: TComponent<P1>) => TComponent<P2>

export const noop: (...args) => any = () => {};

export const idx = (obj, accessor) => {
  try {
    return accessor(obj);
  } catch (error) {
    return undefined;
  }
};

export const isClient = typeof window === 'object';

export const on = (obj, ...args) => obj.addEventListener(...args);

export const off = (obj, ...args) => obj.removeEventListener(...args);

export const ns = (name) => `@@libreact/${name}`;

const hasSymbols = typeof Symbol !== 'undefined';

export const sym = (name) => {
  return hasSymbols ? Symbol(name) : ns(name);
};

export const isFn = (fn) => typeof fn === 'function';

export const h = React.createElement;
