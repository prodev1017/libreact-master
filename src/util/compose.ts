const compose = (...functions) => {
  if (!functions.length) {
    return arg => arg;
  }

  if (functions.length === 1) {
    return functions[0];
  }

  return (...args2) => {
    let result = functions[functions.length - 1](...args2);

    for (let i = functions.length - 2; i >= 0; i--) {
      result = functions[i](result);
    }

    return result;
  };
};

export default compose;
