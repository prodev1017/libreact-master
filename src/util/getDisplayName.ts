const getDisplayName = (Comp: any) => {
  const tipof = typeof Comp;

  switch (tipof) {
    case 'string':
      return Comp;
    case 'function':
      return Comp.displayName || Comp.name || 'Unknown';
    case 'object':
      const {type} = Comp;

      return typeof type === 'function' ?
        `<${type.displayName || type.name || String(Comp)}>` :
        `<Unknown>`;
    default:
      return 'Unknown';
  }
}

export default getDisplayName;
