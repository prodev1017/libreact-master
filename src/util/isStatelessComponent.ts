const isStatelessComponent = (Comp) => {
  if (typeof Comp !== 'function') {
    return false;
  }

  return !Comp.prototype || (Comp instanceof Function);
};

export default isStatelessComponent;
