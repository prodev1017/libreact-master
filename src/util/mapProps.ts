import {h} from '../util';

export const mapProps = (mapper) => (Comp) => (props) => h(Comp, mapper(props));
