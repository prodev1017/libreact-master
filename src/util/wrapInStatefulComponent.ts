import * as React from 'react';
import {h} from '../util';

const wrapInStatefulComponent = (Comp) => {
  const Decorated = class Decorated extends React.Component<any, any> {
    render () {
      return h(Comp, this.props);
    }
  };

  return Decorated;
};

export default wrapInStatefulComponent;
