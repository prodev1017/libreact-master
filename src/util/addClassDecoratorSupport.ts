import isStatelessComponent from './isStatelessComponent';
import wrapInStatefulComponent from './wrapInStatefulComponent';

const addClassDecoratorSupport = (Comp) => {
  return isStatelessComponent(Comp) ?
    Comp :
    wrapInStatefulComponent(Comp);
};

export default addClassDecoratorSupport;
