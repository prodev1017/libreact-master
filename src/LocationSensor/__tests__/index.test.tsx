import {mount} from 'enzyme';
import {h} from '../../util';
import {LocationSensor} from '..';

declare const jsdom;

describe('<LocationSensor>', () => {
  beforeEach(() => {
    jsdom.reconfigure({
      windowTop: window,
      url: "https://example.com/"
    });
  });

  it('is a component', () => {
    expect(LocationSensor).toBeInstanceOf(Function);
  });

  it('renders without crashing', () => {
    mount(
      <LocationSensor>{() =>
        null
      }</LocationSensor>
    );
  });

  it('returns expected default state', () => {
    mount(
      <LocationSensor>{(location) => {
        expect(location).toMatchSnapshot();

        return null;
      }}</LocationSensor>
    );
  });

  describe('re-renders on', () => {
    const events = [
      'pushstate',
      'replacestate',
      'popstate'
    ];

    events.forEach((eventName) => {
      it(eventName, () => {
        const paths = [];
        const triggers = [];

        const wrapper = mount(
          <LocationSensor>{(location) => {
            triggers.push(location.trigger);
            paths.push(location.pathname);

            return null;
          }}</LocationSensor>
        );

        jsdom.reconfigure({
          windowTop: window,
          url: "https://example.com/foobar"
        });

        const event = new Event(eventName);

        window.dispatchEvent(event);

        wrapper.update();

        expect(triggers).toEqual(['load', eventName]);
        expect(paths).toEqual(['/', '/foobar']);
      });
    });
  });
});
