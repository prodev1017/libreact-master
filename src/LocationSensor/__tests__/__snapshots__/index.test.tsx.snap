// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationSensor> re-renders on popstate 1`] = `
Object {
  "hash": "",
  "host": "example.com",
  "hostname": "example.com",
  "href": "https://example.com/foobar",
  "length": 1,
  "origin": "https://example.com",
  "pathname": "/foobar",
  "port": "",
  "protocol": "https:",
  "search": "",
  "state": null,
  "trigger": "popstate",
}
`;

exports[`<LocationSensor> re-renders on pushstate 1`] = `
Object {
  "hash": "",
  "host": "example.com",
  "hostname": "example.com",
  "href": "https://example.com/foobar",
  "length": 1,
  "origin": "https://example.com",
  "pathname": "/foobar",
  "port": "",
  "protocol": "https:",
  "search": "",
  "state": null,
  "trigger": "pushstate",
}
`;

exports[`<LocationSensor> re-renders on replacestate 1`] = `
Object {
  "hash": "",
  "host": "example.com",
  "hostname": "example.com",
  "href": "https://example.com/foobar",
  "length": 1,
  "origin": "https://example.com",
  "pathname": "/foobar",
  "port": "",
  "protocol": "https:",
  "search": "",
  "state": null,
  "trigger": "replacestate",
}
`;

exports[`<LocationSensor> returns expected default state 1`] = `
Object {
  "hash": "",
  "host": "example.com",
  "hostname": "example.com",
  "href": "https://example.com/",
  "length": 1,
  "origin": "https://example.com",
  "pathname": "/",
  "port": "",
  "protocol": "https:",
  "search": "",
  "state": null,
  "trigger": "load",
}
`;
